Added initial version of an integration test

yulric opened the pull request:
the code for this test was taken from the huiport project, in particular the
[create-study-data.R](https://github.com/Big-Life-Lab/huiport/blob/master/R/create-study-data.R) file.

THE TEST IS NOT READY AND SHOULD NOT BE MERGED TO MAIN THE WAY IT IS

The code in this commit is to provide a starting point for an integration test but it has a lot of issues:

1. It uses an external package called cchsflow
2. It includes a lot of extranrous code whereas ideally it would be just a call to `rec_with_table`
3. It may not test everything.

The code should obviously be improved

yulric committed:
added initial version of an integration test

the code for this test was taken from the huiport project, in particular
the
[create-study-data.R](https://github.com/Big-Life-Lab/huiport/blob/master/R/create-study-data.R)
file.

THE TEST IS NOT READY AND SHOULD NOT BE MERGED TO MAIN THE WAY IT IS

The code in this commit is to provide a starting point for an
integration test but it has a lot of issues:

1. It uses an external package called cchsflow
2. It includes a lot of extranrous code whereas ideally it would be just
   a call to `rec_with_table`
3. It may not test everything.

The code should obviously be improved

zargot commented:
thanks, I'll take a look. Do we need the Rdata files?

yulric commented:
> thanks, I'll take a look. Do we need the Rdata files?

Yes, they have the initial data used in the function, https://github.com/Big-Life-Lab/recodeflow/pull/64/files#diff-19034647652e96648dda24526ca68228938c8989b704c0c7e72ca27a9b0d4cf7R105-R113.

zargot committed:
fmt whitespace

zargot committed:
embed cchsflow-v2.1.0 test functions

zargot commented:
@yulric, it's working now. Should I still refactor the code? It may take a long time.

zargot committed:
use snapshot for integration test

yulric committed:
refactored the integration test to remove extraneous code

moved the .RData files into the data folder so that the data can be
programmatically loaded and used without the need for the config.
removed SurveyCycle exception code by removing the variable itself from
the sheet.

yulric commented:
@zargot I've removed all the extraneous code from the test. Can you take a look and merge it in if all looks good.

yulric committed:
switched to using RData for the snapshot so that the attributes can also be tested

zargot committed:
removed sine variables from the integration test

there were OS level differences that was causing the snapshots to be
different.

zargot committed:
update snapshot

zargot commented:
waiting for fix that removes the hui_arc_sine variable (instead of rounding it)

yulric committed:
display snapshot comparison when integration test fails

yulric commented:
@zargot I've updated the tests and snapshot. Can you take a look?

zargot commented:
> @zargot I've updated the tests and snapshot. Can you take a look?

Oh that's neat, thanks! I'll check it, rebase and merge

DougManuel commented on tests/testthat/test-integration.R:
```diff
@@ -0,0 +1,116 @@
+    )
+    return(tidyr::replace_na(x, "NA(c)"))
+  }
+ variables_sheet <- read.csv("./integration-assets/variables.csv", fileEncoding = "UTF-8-BOM")
```
I suggest that we use "UTF-8" without BOM. Re: discussion about git merge errors with our CSV files. I note that if we don't use BOM, we should validate and warn and still may want to accomdate "UFT-8-BOM".

Regardless, we should specify the format somewhere, which is a suggested purpose of PR https://github.com/Big-Life-Lab/recodeflow/pull/70.

Note in the proposed variables.yaml.

```
# CSV file format specification
csv_format:
  encoding: "UTF-8"              # Character encoding
  bom: false                     # No UTF-8 BOM (Byte Order Mark)
  delimiter: ","                 # Field delimiter
  quote_char: '"'                # Character used to quote fields
  escape_char: '"'               # Escape character (doubled quotes)
  line_terminator: "\n"          # Unix-style line endings (LF)
  header_required: true          # First row must contain column names
  header_case_sensitive: true    # Column names must match exactly
  quote_when_needed: true        # Quote fields containing delimiter, quotes, or newlines
  trailing_delimiter: false      # No comma after last field
  blank_lines: "skip"            # Skip blank lines
  comment_char: null             # No comment lines allowed
```

zargot replied:
I agree, we shouldn't use BOM. Should I go ahead and change that, @yulric?

yulric replied:
I'm theoretically good with that but I'm reading online that excel cannot handle CSV files without BOM. That may not be up to date but we should check to make sure. I can do that tomorrow.

yulric replied:
Ok, everything looks good, I was able to open a UTF-8 encoding file with no problems. You can make that change @zargot .

zargot replied:
done

DougManuel replied:
We'll still want to validate and clean CSV files with BOM to help with merge conflicts. Maybe a Git Action.

zargot replied:
@DougManuel, I think it's fine to leave it like this since BOM should be a legacy thing (not the default for new files, etc), but if it becomes an issue in practice then we can reconsider it of course. Is that ok?

DougManuel commented on tests/testthat/integration-assets/cchsflow-functions.R:
```diff
@@ -0,0 +1,487 @@
+  function(DHH_SEX, ALW_1, ALW_2A1, ALW_2A2, ALW_2A3, ALW_2A4, ALW_2A5, ALW_2A6,
+           ALW_2A7) {
+    # If respondents had alcohol in the last week
+    if_else2(ALW_1 == 1,
```
I've noted a more extensive use of `dyplyr` in this PR, which makes sense. We previously had concerns or debate about baseR versus tidyverse. It looks like that debate is over. 😅 https://recode.discourse.group/t/consider-dplyr-dependency-for-cchsflow-and-recodeflow-1-0/36/3

Note that we should declare dplyr as >v1.0, and we'll need to refactor for breaking changes and warnings for this version. 

Should we start depreciating `if_else2()` to dplyr::if_else`?

zargot replied:
this "cchsflow-functions" code is just copy-pasted from cchsflow to avoid having it as a dependency, to further avoid a cyclic dependency in the future (since cchsflow will depend on recodeflow, if I'm not mistaken). It's only used for the test, and I don't think we should spend time on refactoring that code, but cchsflow itself can of course be improved. What do you think?

DougManuel replied:
Agree. It could be any example that avoids cycles. I suggest we use dplyr 1.0+ moving forward when updating the derived functions in packages like `cchsflow`.

yulric commented:
@zargot Can you remove the CCHS .RData files that are not in the data directory?

zargot committed:
remove old rdata files

zargot committed:
fix rdata file mode

zargot commented:
> @zargot Can you remove the CCHS .RData files that are not in the data directory?

done

zargot committed:
stop using BOM (with UTF-8)

to adhere to best practice.

The CSV files themselves don't actually have BOM since they're plain
ASCII, so no change is needed for them.

yulric commented:
@zargot You can merge this in whenever, I can't give an approval since I'm the one who made the PR.

zargot committed:
fix tidyselect warning

"Using an external vector in selections was deprecated in tidyselect
1.1.0."
